runtime/flashbox/src/lib.rs

```diff
@@ -0,0 +1,1535 @@
+    // Essentially just Aura, but let's be pedantic.
+    type SessionHandler = <SessionKeys as sp_runtime::traits::OpaqueKeys>::KeyTypeIdProviders;
+    type Keys = SessionKeys;
+    type WeightInfo = pallet_session::weights::SubstrateWeight<Runtime>;
+}
+
+pub struct RemoveInvulnerablesImpl;
+
+impl RemoveInvulnerables<AccountId> for RemoveInvulnerablesImpl {
+    fn remove_invulnerables(
+        collators: &mut Vec<AccountId>,
+        num_invulnerables: usize,
+    ) -> Vec<AccountId> {
+        if num_invulnerables == 0 {
+            return vec![];
+        }
+        // TODO: check if this works on session changes
+        let all_invulnerables = pallet_invulnerables::Invulnerables::<Runtime>::get();
+        if all_invulnerables.is_empty() {
+            return vec![];
+        }
+        let mut invulnerables = vec![];
+        // TODO: use binary_search when invulnerables are sorted
+        collators.retain(|x| {
+            if invulnerables.len() < num_invulnerables && all_invulnerables.contains(x) {
+                invulnerables.push(x.clone());
+                false
+            } else {
+                true
+            }
+        });
+
+        invulnerables
+    }
+}
+
+pub struct RemoveParaIdsWithNoCreditsImpl;
+
+impl RemoveParaIdsWithNoCredits for RemoveParaIdsWithNoCreditsImpl {
+    fn remove_para_ids_with_no_credits(para_ids: &mut Vec<ParaId>) {
+        let blocks_per_session = Period::get();
+        let credits_for_2_sessions = 2 * blocks_per_session;
+        para_ids.retain(|para_id| {
+            // Check if the container chain has enough credits for producing blocks for 2 sessions
+            let credits = pallet_services_payment::BlockProductionCredits::<Runtime>::get(para_id)
+                .unwrap_or_default();
+
+            credits >= credits_for_2_sessions
+        });
+    }
+
+    /// Make those para ids valid by giving them enough credits, for benchmarking.
+    #[cfg(feature = "runtime-benchmarks")]
+    fn make_valid_para_ids(para_ids: &[ParaId]) {
+        use frame_support::assert_ok;
+
+        let blocks_per_session = Period::get();
+        // Enough credits to run any benchmark
+        let credits = 20 * blocks_per_session;
+
+        for para_id in para_ids {
+            assert_ok!(ServicesPayment::set_credits(
+                RuntimeOrigin::root(),
+                *para_id,
+                credits,
+            ));
+        }
+    }
+}
+
+pub struct NeverRotateCollators;
+
+impl ShouldRotateAllCollators<u32> for NeverRotateCollators {
+    fn should_rotate_all_collators(_: u32) -> bool {
+        false
+    }
+}
+
+impl pallet_collator_assignment::Config for Runtime {
+    type RuntimeEvent = RuntimeEvent;
+    type HostConfiguration = Configuration;
+    type ContainerChains = Registrar;
+    type SessionIndex = u32;
+    type SelfParaId = ParachainInfo;
+    type ShouldRotateAllCollators = NeverRotateCollators;
+    type GetRandomnessForNextBlock = ();
```

Comment:
Since flashbox doesn't use randomness, container chains with a lower para id will be given priority when assigning collators.

Reply:
yes, that is fine, it will only happen if we dont have sufficient collators to serve demand